hints: make hints concurrency configurable and reduce the default

piodul opened the pull request:
Previously, hinted handoff had a hardcoded concurrency limit - at most
128 hints could be sent from a single shard at once. This commit makes
this limit configurable by adding a new configuration option:
`max_hinted_handoff_concurrency_per_shard`. This option can be updated
in runtime. Additionally, the default concurrency per shard is made
lower and is now 8.

The motivation for reducing the concurrency was to mitigate the negative
impact hints may have on performance of the receiving node due to them
not being properly isolated with respect to I/O.

Tests:
- unit(dev)
- dtests(hintedhandoff_additional_test.py)

Refs: #8624

piodul commented:
Few issues/questions from my side:

- We already have an (unused) `max_hints_delivery_threads` option - what should we do with it? Re-use, delete? I think having them both might be confusing.
- Should the concurrency limit be per-shard, or maybe per-node? Should the limit scale down with the number of nodes (so that the receiver concurrency is bound)?
- Is the limit of 32 hints fine? It was a completely arbitrary choice - should we make it lower/higher? Or maybe run a test of some sort?

piodul commented:
In the meantime, I'm running hinted handoff dtests (hintedhandoff_additional_test.py), I'll update information about the tests in the PR description/commit description after they finish (or after I fix potential bugs).

haaawk commented on db/hints/resource_manager.cc:
```diff
@@ -52,7 +52,8 @@ future<bool> is_mountpoint(const fs::path& path) {
 future<semaphore_units<named_semaphore::exception_factory>> resource_manager::get_send_units_for(size_t buf_size) {
     // Let's approximate the memory size the mutation is going to consume by the size of its serialized form
-    size_t hint_memory_budget = std::max(_min_send_hint_budget, buf_size);
+    const size_t max_send_hint_budget = _max_send_in_flight_memory / std::max(uint32_t(1), _max_hints_send_queue_length());
```
This is a bit confusing. Previously the variable was called `_min_send_hint_budget` and was used with `std::max` which means it was a lower bound for the budget. Now the variable is called `max_send_hint_budget` which creates an impression that it's an upper bound but it's still used in `std::max` so it's a lower bound.

piodul replied:
My mistake. The variable should be named `min_send_hint_budget`. Apart from dropping the underscore, I accidentally changed min to max - not sure why. I'll fix it after the dtests finish.

piodul replied:
Done, changed to `min_send_hint_budget`. As a bonus, I added a comment which explains why the hint send budget is calculated this way.

haaawk reviewed (approved):
LGTM

avikivity commented:
> Few issues/questions from my side:
> 
> * We already have an (unused) `max_hints_delivery_threads` option - what should we do with it? Re-use, delete? I think having them both might be confusing.

Reusing is possible. But note we need to divide by the number of shards, since the user expects a per-node configuration.

> * Should the concurrency limit be per-shard, or maybe per-node? Should the limit scale down with the number of nodes (so that the receiver concurrency is bound)?
> * Is the limit of 32 hints fine? It was a completely arbitrary choice - should we make it lower/higher? Or maybe run a test of some sort?

I'm guessing it's still too high. You can give @fgelcer packages with (say) 8 and 32 and see what happens.

> 
> @avikivity @vladzcloudius

vladzcloudius reviewed (approved):
LGTM.

But as I already mentioned, while this patch doesn't hurt I doubt it is going to resolve any of the issues at hand that eventually cause the receiver side of hints to get degraded performance during the intensive hints sending (e.g. after node restart):
 * Too high concurrency on the hints receiver side.
 * Too high throughput of hints on the receiver side.

Each of these can only be handled on the receiver side of hints.

asias commented:
> > Few issues/questions from my side:
> > 
> > * We already have an (unused) `max_hints_delivery_threads` option - what should we do with it? Re-use, delete? I think having them both might be confusing.
> 
> Reusing is possible. But note we need to divide by the number of shards, since the user expects a per-node configuration.

If we do not use the max_hints_delivery_threads yet, I guess it is better to use the new and better option:  max_hinted_handoff_concurrency_per_shard.

The threads part in that name does not make any sense to scylla.

From the experience and feedback of reusing the seed config option, it would be much cleaner to use a new option --contact-points instead. 

> 
> > * Should the concurrency limit be per-shard, or maybe per-node? Should the limit scale down with the number of nodes (so that the receiver concurrency is bound)?
> > * Is the limit of 32 hints fine? It was a completely arbitrary choice - should we make it lower/higher? Or maybe run a test of some sort?
> 
> I'm guessing it's still too high. You can give @fgelcer packages with (say) 8 and 32 and see what happens.

+1. Since hints replay is of lower priority. It is better to be more conservative. I would vote for 8 or even 4. 

> 
> > @avikivity @vladzcloudius

piodul commented:
v2:
- Fixed a bug caused by cross-shard use of updateable_value (https://github.com/scylladb/scylla/issues/7316 strikes again...)
- Changed the default concurrency from 32 to 8

I changed the default to a lower value already, but I am still going to post performance results here after I get them so we may make a more informed decision.

avikivity commented on db/config.cc:
```diff
@@ -591,6 +591,8 @@ db::config::config(std::shared_ptr<db::extensions> exts)
         "Enable or disable hinted handoff. To enable per data center, add data center list. For example: hinted_handoff_enabled: DC1,DC2. A hint indicates that the write needs to be replayed to an unavailable node. "
         "Related information: About hinted handoff writes")
+    , max_hinted_handoff_concurrency_per_shard(this, "max_hinted_handoff_concurrency_per_shard", liveness::LiveUpdate, value_status::Used, 8,
+        "Maximum concurrency allowed for sending hints from a single shard. Setting this value to 0 does not disable hinted handoff, instead concurrency 1 will be used.")
```
Better to make it per-node. Users don't expect per-shard configuration. The default can be 8*smp::count (or even the 4 that @vladzcloudius prefers).

vladzcloudius replied:
> or even the 4 that @vladzcloudius prefers
@avikivity, are you referring my `min_threshold` preference? ;)

If we make it per-node what shell we do if the user chooses a value less than the number of shards? Not multiple of shards? Fail to boot? 
If we choose to have a per-node value then the concurrency controller would rather be per-node too and this is going to be quite against a seastar mindset.

As unexpected as it is customers can deal with a per-shard parameter just alright.

piodul replied:
I'd vote for a per-shard parameter, too. I don't see how we can enforce a per-node limit without doing cross-shard operations. Besides, we already have a per-shard parameter in the configuration - `max_concurrent_requests_per_shard`.

avikivity replied:
@piodul divide the concurrency by the number of shards (with std::max(1, ...))  and that's your concurrency limit.

We have a similar arrangement with commitlog size.

piodul replied:
@avikivity What about numbers not divisible by the shard count? Commitlog size is a number expressed in megabytes, so rounding down or up doesn't hurt. If somebody asks for a very low hint concurrency on the node (e.g. between `2*shard_count` or `3*shard_count`), rounding will have significant effect.

avikivity replied:
It's okay if it's imperfect. I'm trying to avoid a situation where some config variables are per node, some per shard.

piodul replied:
OK. I changed it to a per-node parameter. Concurrency is now rounded up (as in the case of the commitlog size).

piodul commented:
I ran the performance regression job with 32 and 8 hints per shard, and here are the graphs.

32 hints per shard:

![parallelism-32](https://user-images.githubusercontent.com/4950936/118676561-cd103d80-b7fb-11eb-85ca-b3fb6db7783f.png)

8 hints per shard:

![parallelism-8](https://user-images.githubusercontent.com/4950936/118676580-d0a3c480-b7fb-11eb-930e-6f1f23ac99e9.png)

We can see that the feature which waits for hints to be sent before repair delayed repair a little, but unfortunately it seems it gave up before all hints were sent (currently it will give up waiting for a particular hint queue if a hint failed to be sent, or won't keep waiting for a node if it failed to respond to the HINT_SYNC_POINT_CHECK). Here it might actually be fortunate for us because we can observe how the latency behaves when both hints and repair are running.

It looks like 8 hints per shard is a better choice. Looking at the p95 graph, I can't see the difference in latency between hints+repair vs. repair alone. Hints+regular writes vs. writes alone also doesn't look tragic. For 32 hints per shard, the difference is visible.

As for waiting for hints to be sent... I'll try to harden it as a part of converting it to HTTP API.

piodul commented:
More details about the test runs can be found in Jenkins (thanks @fgelcer for the job!):

- 8 hints/shard: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/fabio/job/hh_parallelism/job/perf-latency-regression-hh-parallel/11/
- 32 hints/shard: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/fabio/job/hh_parallelism/job/perf-latency-regression-hh-parallel/9/

piodul commented:
@asias @avikivity @vladzcloudius Could you look at the results above? Is concurrency 8 OK? Is there anything you would like me to do here?

vladzcloudius commented:
> @asias @avikivity @vladzcloudius Could you look at the results above? Is concurrency 8 OK? Is there anything you would like me to do here?

@piodul it seems like 8 was fine in this test case.
But note that latency goes high on the hints receiver side nodes. 
This means that if they get higher concurrently, e.g. because there are more "other" nodes - 8 may start looking like 32 above. Even 1 may start looking like 32 or even worse.

Like I already mentioned - I don't think that this patch is going to resolve anything.
But I agree that it doesn't hurt. - It's an easy tool that may help sometimes. 
However we need to start working on properly limiting the incoming concurrency on the receiver side.

asias commented:
> We can see that the feature which waits for hints to be sent before repair delayed repair a little, but unfortunately it seems it gave up before all hints were sent (currently it will give up waiting for a particular hint queue if a hint failed to be sent, or won't keep waiting for a node if it failed to respond to the HINT_SYNC_POINT_CHECK). Here it might actually be fortunate for us because we can observe how the latency behaves when both hints and repair are running.

Thanks @piodul. Do you know the exact reason why the waiting finishes before hints are fully replayed. I think there was no node down during your tests. If this is the case, why would the reply fail in the first place? Is the waiting time are too short?

> It looks like 8 hints per shard is a better choice. Looking at the p95 graph, I can't see the difference in latency between hints+repair vs. repair alone. Hints+regular writes vs. writes alone also doesn't look tragic. For 32 hints per shard, the difference is visible.

Yes. 

> As for waiting for hints to be sent... I'll try to harden it as a part of converting it to HTTP API.

OK.

piodul commented:
> > We can see that the feature which waits for hints to be sent before repair delayed repair a little, but unfortunately it seems it gave up before all hints were sent (currently it will give up waiting for a particular hint queue if a hint failed to be sent, or won't keep waiting for a node if it failed to respond to the HINT_SYNC_POINT_CHECK). Here it might actually be fortunate for us because we can observe how the latency behaves when both hints and repair are running.
> 
> Thanks @piodul. Do you know the exact reason why the waiting finishes before hints are fully replayed. I think there was no node down during your tests. If this is the case, why would the reply fail in the first place? Is the waiting time are too short?
> 

@asias I can't tell for sure, but my theory is that a hint failed due to a timeout. Hints are sent using the same framework as replica writes, and they use the same timeout. When a hint times out, we consider it as a failure and do not wait for this hint queue any longer.

Perhaps we should make waiting code not stop when a hint times out, and stop waiting only if one of the nodes is considered down.

asias commented:
> > > We can see that the feature which waits for hints to be sent before repair delayed repair a little, but unfortunately it seems it gave up before all hints were sent (currently it will give up waiting for a particular hint queue if a hint failed to be sent, or won't keep waiting for a node if it failed to respond to the HINT_SYNC_POINT_CHECK). Here it might actually be fortunate for us because we can observe how the latency behaves when both hints and repair are running.
> > 
> > 
> > Thanks @piodul. Do you know the exact reason why the waiting finishes before hints are fully replayed. I think there was no node down during your tests. If this is the case, why would the reply fail in the first place? Is the waiting time are too short?
> 
> @asias I can't tell for sure, but my theory is that a hint failed due to a timeout. Hints are sent using the same framework as replica writes, and they use the same timeout. When a hint times out, we consider it as a failure and do not wait for this hint queue any longer.
> 
> Perhaps we should make waiting code not stop when a hint times out, and stop waiting only if one of the nodes is considered down.

Hint replay is a maintenance operation. Such per single hint timeout is not very useful. Unlike user writes, we do not care much if the replay takes longer since it is has lower priority anyway.  It is bad such failure would cause the whole replay operation to stop. I guess we can make the timeout much longer or even do not use it. What is the timeout, right now?

Btw, since we are moving the control of replay to REST api, it would be good to have an api to abort the replay.

piodul commented:
v3: changed the `max_concurrent_requests_per_shard` parameter to a per-node parameter - `max_concurrent_requests`. The concurrency is divided across shards and rounded up if it does not divide the number of shards evenly. If not provided, concurrency of 8 per shard will be used.

piodul commented:
@avikivity ping

piodul commented:
@avikivity ping^2

avikivity commented:
Sorry, reviewing.

avikivity commented:
@fgelcer let's see if this helps with latency while node operations are done.

fgelcer commented:
> @fgelcer let's see if this helps with latency while node operations are done.

that are great news!!!
we do have some results from this week's trigger (from June 25th) and they do look fine, but mixed workload that is slightly above the 5 ms threshold
